text/0000-prepare-operation.md

```diff
@@ -0,0 +1,387 @@
+  - **Goal 1: Serializing CLI configuration**
+    > As a user, I would like to be able to serialize and share the parameters I use with certain platforms such as `pack`.
+  - **Goal 2: A recognizable file in repositories**
+    > As a user, I would like to be able to recognize, based on the file system, if a project is using Cloud Native Buildpacks.
+  - **Goal 3: Platform recognition of project.toml**
+    > As a user, I would like to ensure that my configuration in `project.toml` is being used.
+
+**This RFC directly solves for Goal 3**, while enabling the solutions for Goals 1 and 2 via a supplemental RFC ([Support for pack.toml][rfc-pr-189]).
+
+[rfc-pr-189]: https://github.com/buildpacks/rfcs/pull/189
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is composed of following changes:
+
+- [Changes to `io.buildpacks` namespace.](#changes-to-iobuildpacks-namespace)
+  - [Define criteria for properties in namespace.](#criteria-for-properties)
+  - [Version `io.buildpacks` namespace.](#version-iobuildpacks-namespace)
+- [A replaceable new phase `prepare` that applies platform configuration.](#prepare-phase)
+- [Supporting Cloud Native Buildpack utilities.](#cloud-native-buildpacks-utilities)
+
+## Changes to `io.buildpacks` namespace
+
+### Criteria for properties
+
+This namespace declares default values that platforms should acknowledge. A majority of properties should be spec'd but we would allow for unspec'd additional properties.
+
+The criteria for spec'd properties in this namespace would be based on whether they affect inputs to the lifecycle phases. For example, the order and environment variables are inputs to the lifecycle. The `builder`, in contrast, is not an input to the lifecycle so it wouldn't meet the criteria.
+
+Therefore the following changes have been made from the [latest schema][pd-02]:
+
+ - **Remove `builder` from `io.buildpacks` namespace**
+
+[pd-02]: https://github.com/buildpacks/spec/blob/extensions/project-descriptor/0.2/extensions/project-descriptor.md#iobuildpacks-optional
+
+### Version `io.buildpacks` namespace
+
+The `io.buildpacks` namespace schema should be able to change without changing the Project Descriptor spec. For this reason, the namespace should define a `schema-version`.
+
+### Schema
+
+See [`io.buildpacks` Namespace Schema](#io-buildpacks-Namespace-Schema) section below.
+
+#### Usages Examples
+
+###### Multi-Platform (pack + kpack)
+
+In the following example we show how various platform namespaces can have their own configuration but properties can be promoted to `io.buildpacks` if the user's intent is to apply it to all platforms.
+
+```toml
+[io.buildpacks]
+schema-version = "0.2"
+
+###
+# common buildpacks config
+##
+
+[io.buildpacks]
+exclude = ["some-files/**"]
+
+##
+# pack config
+##
+
+[io.buildpacks.pack]
+schema-version = "0.1"
+builder = "cnbs/sample-builder:bionic"
+cache-volume = "my-app-cache"           # (example only)
+
+##
+# kpack config
+##
+
+[com.vmware.kpack]
```

Comment:
the `com.vmware` is a bit problematic. Does kpack have a vendor neutral namespace?

Reply:
I'm guessing that will depend on #235, but I'll look into it